internal/decode/decode.go

```diff
@@ -0,0 +1,153 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package decode implements a generic interface{} decoder. The intention is
+// to use it to decode arbitrary map[interface{}]interface{} objects into
+// structs or other complex objects.
+package decode
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+const _tagName = "config"
+
+var _typeOfDecoder = reflect.TypeOf((*Decoder)(nil)).Elem()
+
+// Into is a function that attempts to decode the source data into the given
+// destination. dst MUST be a pointer to a value.
+//
+// 	var (
+// 		decode decode.Into = ...
+// 		value map[string]MyStruct
+// 	)
+// 	err := decode(&value)
+type Into func(dst interface{}) error
+
+// Decoder is any type which has custom decoding logic. Types may implement
+// Decode and rely on the given Decode function to read values.
+//
+// For example,
+//
+// 	type StringSet map[string]struct{}
+//
+// 	func (ss *StringSet) Decode(dec decode.Into) error {
+// 		var items []string
+// 		if err := dec(&items); err != nil {
+// 			return err
+// 		}
+//
+// 		*ss = make(map[string]struct{})
+// 		for _, item := range items {
+// 			(*ss)[item] = struct{}{}
+// 		}
+// 		return nil
+// 	}
+type Decoder interface {
+	// Decode receives a function that will attempt to decode the source data
+	// into the given target. The argument MUST be a pointer to the target
+	// object.
+	Decode(Into) error
+}
+
+// Decode from src into dst. dst may implement Decoder to customize how src is
+// read into it.
+func Decode(dst, src interface{}) error {
+	return decodeFrom(src)(dst)
+}
+
+// decodeFrom builds a decode Into function that reads the given value into
+// the destination.
+func decodeFrom(src interface{}) Into {
+	return func(dst interface{}) error {
+		cfg := mapstructure.DecoderConfig{
+			ErrorUnused: true,
+			Result:      dst,
+			DecodeHook: mapstructure.ComposeDecodeHookFunc(
+				mapstructure.StringToTimeDurationHookFunc(),
+				decoderDecodeHook,
+			),
+			TagName: _tagName,
+		}
+
+		dec, err := mapstructure.NewDecoder(&cfg)
```

Comment:
Suggest `decoder` here, since it's only called once below and I'd rather not have to think.